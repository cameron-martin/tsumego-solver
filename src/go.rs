mod benson;
mod bit_board;
mod sgf_conversion;
pub use bit_board::{BitBoard, BoardPosition};
use std::collections::hash_map::DefaultHasher;
use std::fmt;
use std::fmt::Debug;
use std::fmt::Formatter;
use std::fmt::{Display, Write};
use std::hash::{Hash, Hasher};

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Move {
    PassTwice,
    PassOnce,
    Place(BoardPosition),
}

impl Display for Move {
    fn fmt(&self, f: &mut Formatter) -> Result<(), fmt::Error> {
        match self {
            Move::PassOnce => f.write_str("Pass (once)"),
            Move::PassTwice => f.write_str("Pass (twice)"),
            Move::Place(position) => f.write_fmt(format_args!("{}", position)),
        }
    }
}

#[derive(Copy, Clone, Debug, PartialEq)]
pub enum BoardCell {
    Empty,
    OutOfBounds,
    Occupied(GoPlayer),
}

#[derive(Copy, Clone, Debug, PartialEq)]
pub enum GoPlayer {
    Black,
    White,
}

impl GoPlayer {
    pub fn flip(self) -> GoPlayer {
        match self {
            GoPlayer::Black => GoPlayer::White,
            GoPlayer::White => GoPlayer::Black,
        }
    }

    pub fn both() -> impl Iterator<Item = &'static GoPlayer> {
        [GoPlayer::Black, GoPlayer::White].iter()
    }
}

impl Display for GoPlayer {
    fn fmt(&self, f: &mut Formatter) -> Result<(), fmt::Error> {
        match self {
            GoPlayer::Black => f.write_str("black"),
            GoPlayer::White => f.write_str("white"),
        }
    }
}

// Being set in both black and white denotes "out of bounds"
#[derive(PartialEq, Clone, Copy, Debug, Hash)]
pub struct GoBoard {
    white: BitBoard,
    black: BitBoard,
}

impl Display for GoBoard {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        for j in 0..BitBoard::height() {
            for i in 0..BitBoard::width() {
                if i != 0 {
                    f.write_char(' ')?;
                }

                f.write_char(match self.get_cell(BoardPosition::new(i, j)) {
                    BoardCell::Empty => '.',
                    BoardCell::OutOfBounds => '_',
                    BoardCell::Occupied(GoPlayer::White) => 'w',
                    BoardCell::Occupied(GoPlayer::Black) => 'b',
                })?;
            }
            f.write_str("\n")?;
        }

        Ok(())
    }
}

impl GoBoard {
    fn empty() -> GoBoard {
        GoBoard {
            white: BitBoard::empty(),
            black: BitBoard::empty(),
        }
    }

    pub fn new(black: BitBoard, white: BitBoard, out_of_bounds: BitBoard) -> GoBoard {
        GoBoard {
            white: white | out_of_bounds,
            black: black | out_of_bounds,
        }
    }

    /// Empty cells, including out of bounds
    fn empty_cells(&self) -> BitBoard {
        !(self.white ^ self.black)
    }

    pub fn out_of_bounds(&self) -> BitBoard {
        self.white & self.black
    }

    fn set_cell(&mut self, position: BoardPosition, cell: BoardCell) {
        let mask = BitBoard::singleton(position);

        match cell {
            BoardCell::Empty => {
                self.black = self.black & !mask;
                self.white = self.white & !mask;
            }
            BoardCell::Occupied(GoPlayer::Black) => {
                self.black = self.black | mask;
                self.white = self.white & !mask;
            }
            BoardCell::Occupied(GoPlayer::White) => {
                self.black = self.black & !mask;
                self.white = self.white | mask;
            }
            BoardCell::OutOfBounds => {
                self.black = self.black | mask;
                self.white = self.white | mask;
            }
        }
    }

    fn get_cell(&self, position: BoardPosition) -> BoardCell {
        let mask = BitBoard::singleton(position);

        if !((mask & self.out_of_bounds()).is_empty()) {
            return BoardCell::OutOfBounds;
        }

        if !((mask & self.white).is_empty()) {
            return BoardCell::Occupied(GoPlayer::White);
        }

        if !((mask & self.black).is_empty()) {
            return BoardCell::Occupied(GoPlayer::Black);
        }

        BoardCell::Empty
    }

    pub fn get_bitboard_for_player(&self, player: GoPlayer) -> BitBoard {
        match player {
            GoPlayer::Black => (self.black & !self.white),
            GoPlayer::White => (self.white & !self.black),
        }
    }

    fn set_bitboard_for_player(&mut self, player: GoPlayer, board: BitBoard) {
        match player {
            GoPlayer::Black => self.black = board | self.out_of_bounds(),
            GoPlayer::White => self.white = board | self.out_of_bounds(),
        }
    }

    fn get_bitboard_at_position(&self, position: BoardPosition) -> BitBoard {
        let mask = BitBoard::singleton(position);
        let white = self.get_bitboard_for_player(GoPlayer::White);
        let black = self.get_bitboard_for_player(GoPlayer::Black);

        if !((mask & white).is_empty()) {
            return white;
        }

        if !((mask & black).is_empty()) {
            return black;
        }

        panic!("No board at this position")
    }

    fn group_has_liberties(&self, position: BoardPosition) -> bool {
        let mask = self.get_bitboard_at_position(position);

        let group = BitBoard::singleton(position).flood_fill(mask);

        !(group.expand_one() & self.empty_cells()).is_empty()
    }

    fn get_alive_groups_for_player(&self, player: GoPlayer) -> BitBoard {
        let bitboard = self.get_bitboard_for_player(player);

        (self.empty_cells().expand_one() & bitboard).flood_fill(bitboard)
    }

    fn remove_dead_groups_for_player(&mut self, player: GoPlayer) {
        let stones_with_liberties = self.get_alive_groups_for_player(player);

        self.set_bitboard_for_player(player, stones_with_liberties);
    }

    pub fn has_dead_groups(&self) -> bool {
        GoPlayer::both().any(|&player| {
            let alive_groups = self.get_alive_groups_for_player(player);

            self.get_bitboard_for_player(player) != alive_groups
        })
    }

    fn is_out_of_bounds(&self, position: BoardPosition) -> bool {
        !(BitBoard::singleton(position) & self.out_of_bounds()).is_empty()
    }

    fn set_out_of_bounds(&mut self, out_of_bounds: BitBoard) {
        let prev_out_of_bounds = self.out_of_bounds();
        self.white = (self.white & !prev_out_of_bounds) | out_of_bounds;
        self.black = (self.black & !prev_out_of_bounds) | out_of_bounds;
    }

    pub fn stable_hash(&self) -> u64 {
        let mut hasher = DefaultHasher::new();
        self.hash(&mut hasher);

        hasher.finish()
    }
}

#[derive(Debug, PartialEq, Clone, Copy)]
pub struct GoGame {
    ko_violations: BitBoard,
    board: GoBoard,
    pub current_player: GoPlayer,
    pub last_move_pass: bool,
}

#[derive(Debug, PartialEq)]
pub enum MoveError {
    Occupied,
    OutOfBounds,
    OutOfTurn,
    Suicidal,
    Ko,
}

impl GoGame {
    pub fn empty() -> GoGame {
        GoGame {
            board: GoBoard::empty(),
            ko_violations: BitBoard::empty(),
            current_player: GoPlayer::Black,
            last_move_pass: false,
        }
    }

    pub fn from_board(board: GoBoard, current_player: GoPlayer) -> GoGame {
        GoGame {
            board,
            ko_violations: BitBoard::empty(),
            current_player,
            last_move_pass: false,
        }
    }

    pub fn get_board(&self) -> GoBoard {
        self.board
    }

    fn get_cell(&self, position: BoardPosition) -> BoardCell {
        self.get_board().get_cell(position)
    }

    fn is_out_of_bounds(&self, position: BoardPosition) -> bool {
        self.get_board().is_out_of_bounds(position)
    }

    pub fn play_move_for_player(
        &self,
        go_move: Move,
        player: GoPlayer,
    ) -> Result<GoGame, MoveError> {
        if self.current_player != player {
            return Err(MoveError::OutOfTurn);
        }

        self.play_move(go_move)
    }

    pub fn play_move(&self, go_move: Move) -> Result<GoGame, MoveError> {
        match go_move {
            Move::Place(position) => self.play_placing_move(position),
            Move::PassOnce => Ok(self.pass()),
            Move::PassTwice => Ok(self.pass()),
        }
    }

    pub fn play_placing_move(&self, position: BoardPosition) -> Result<GoGame, MoveError> {
        if self.get_cell(position) != BoardCell::Empty {
            return Err(MoveError::Occupied);
        }

        if self.is_out_of_bounds(position) {
            return Err(MoveError::OutOfBounds);
        }

        let next_player = self.current_player.flip();

        let mut new_board = self.get_board();
        new_board.set_cell(position, BoardCell::Occupied(self.current_player));

        // Remove dead groups owned by other player
        new_board.remove_dead_groups_for_player(next_player);

        // Evaluate suicide
        if !new_board.group_has_liberties(position) {
            return Err(MoveError::Suicidal);
        }

        // Evaluate ko
        if self.ko_violations.is_set(position) {
            return Err(MoveError::Ko);
        }

        let ko_violations = if (BitBoard::singleton(position).immediate_exterior()
            & self
                .get_board()
                .get_bitboard_for_player(self.current_player))
        .is_empty()
        {
            (self.board.get_bitboard_for_player(next_player)
                & !new_board.get_bitboard_for_player(next_player))
            .singletons()
        } else {
            BitBoard::empty()
        };

        Ok(GoGame {
            ko_violations,
            board: new_board,
            current_player: next_player,
            last_move_pass: false,
        })
    }

    pub fn pass(&self) -> GoGame {
        GoGame {
            board: self.board,
            ko_violations: BitBoard::empty(),
            current_player: self.current_player.flip(),
            last_move_pass: true,
        }
    }

    pub fn generate_moves(&self) -> Vec<(GoGame, Move)> {
        let mut games = Vec::new();

        let board = self.get_board();

        for position in (!(board.white | board.black)).positions() {
            if let Ok(game) = self.play_placing_move(position) {
                games.push((game, Move::Place(position)));
            }
        }

        games
    }

    pub fn generate_moves_including_pass(&self) -> Vec<(GoGame, Move)> {
        let mut games = self.generate_moves();

        games.push((
            self.pass(),
            if self.last_move_pass {
                Move::PassTwice
            } else {
                Move::PassOnce
            },
        ));

        games
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn can_add_stone() {
        let game = GoGame::empty();
        let game = game
            .play_move_for_player(Move::Place(BoardPosition::new(0, 0)), GoPlayer::Black)
            .unwrap();

        assert_eq!(
            game.get_cell(BoardPosition::new(0, 0)),
            BoardCell::Occupied(GoPlayer::Black)
        );
    }

    #[test]
    fn previous_board_is_not_mutated() {
        let old_game = GoGame::empty();
        let new_game = old_game
            .play_move_for_player(Move::Place(BoardPosition::new(0, 0)), GoPlayer::Black)
            .unwrap();

        assert_ne!(
            old_game.get_cell(BoardPosition::new(0, 0)),
            new_game.get_cell(BoardPosition::new(0, 0))
        );
    }

    #[test]
    fn current_player_starts_as_black() {
        let game = GoGame::empty();

        assert_eq!(game.current_player, GoPlayer::Black);
    }

    #[test]
    fn player_advances_when_playing_move() {
        let game = GoGame::empty()
            .play_placing_move(BoardPosition::new(0, 0))
            .unwrap();

        assert_eq!(game.current_player, GoPlayer::White);
    }

    #[test]
    fn cannot_play_move_out_of_turn() {
        let result = GoGame::empty()
            .play_move_for_player(Move::Place(BoardPosition::new(0, 0)), GoPlayer::White);

        assert_eq!(result, Err(MoveError::OutOfTurn));
    }

    #[test]
    fn cannot_play_in_occupied_space() {
        let game = GoGame::empty()
            .play_placing_move(BoardPosition::new(0, 0))
            .unwrap();
        let result = game.play_placing_move(BoardPosition::new(0, 0));

        assert_eq!(result, Err(MoveError::Occupied));
    }

    #[test]
    fn single_groups_are_captured() {
        let game = GoGame::from_sgf(include_str!("test_sgfs/single_groups_are_captured.sgf"));

        assert_eq!(game.get_cell(BoardPosition::new(0, 0)), BoardCell::Empty);
    }

    #[test]
    fn complex_groups_are_captured() {
        let game = GoGame::from_sgf(include_str!("test_sgfs/complex_capture.sgf"));
        let game = game.play_placing_move(BoardPosition::new(11, 6)).unwrap();

        assert_eq!(
            format!("{}", game.get_board()),
            ". . . . . . . . . . . . . . . .\n\
             . b . b . b b w w . w . . . . .\n\
             . . . . . . . b w w . w . w . .\n\
             . . . b . b b . w . . w b . . .\n\
             . . . . b . . . w w . w . . . .\n\
             . . . b . . w . w . . . w . . .\n\
             . . . . . . . . . w w w . . . .\n\
             . . . . . . . . . . . . . . . .\n"
        );
    }

    #[test]
    fn capturing_has_precedence_over_suicide() {
        let game = GoGame::from_sgf(include_str!(
            "test_sgfs/capturing_has_precedence_over_suicide.sgf"
        ));

        assert_eq!(game.get_cell(BoardPosition::new(1, 0)), BoardCell::Empty);
    }

    #[test]
    fn cannot_commit_suicide() {
        let game = GoGame::from_sgf(include_str!("test_sgfs/cannot_commit_suicide.sgf"));
        let result = game.play_placing_move(BoardPosition::new(0, 0));

        assert_eq!(result, Err(MoveError::Suicidal));
    }

    #[test]
    fn ko_rule_simple() {
        let game = GoGame::from_sgf(include_str!("test_sgfs/ko_rule_simple.sgf"));
        let result = game.play_placing_move(BoardPosition::new(2, 2));

        assert_eq!(result, Err(MoveError::Ko));
    }

    #[test]
    fn capture_two_recapture_one_not_ko_violation() {
        let game = GoGame::from_sgf(include_str!("test_sgfs/capture_two_recapture_one.sgf"));

        game.play_placing_move(BoardPosition::new(3, 2)).unwrap();
    }

    #[test]
    fn capturing_single_and_joining_group_does_not_trigger_ko() {
        let game = GoGame::from_sgf(include_str!("test_sgfs/capture_single_join_group.sgf"));

        let result = game.play_placing_move(BoardPosition::new(2, 1));

        assert!(result.is_ok());
    }

    #[test]
    fn out_of_bounds_moves_are_not_generated() {
        let game = GoGame::from_sgf(include_str!("test_sgfs/puzzles/true_simple1.sgf"));
        let moves = game.generate_moves();

        assert_eq!(moves.len(), 5);
    }

    #[test]
    fn pass_sets_last_move_pass() {
        let game = GoGame::from_sgf(include_str!("test_sgfs/ko_rule_simple.sgf"));
        let new_game = game.pass();

        assert!(new_game.last_move_pass);
    }

    #[test]
    fn move_clears_last_move_pass() {
        let game = GoGame::from_sgf(include_str!("test_sgfs/ko_rule_simple.sgf"));
        let game = game.pass();
        let game = game.play_placing_move(BoardPosition::new(13, 7)).unwrap();

        assert!(!game.last_move_pass);
    }

    #[test]
    fn pass_advances_player() {
        let game = GoGame::from_sgf(include_str!("test_sgfs/ko_rule_simple.sgf"));
        let new_game = game.pass();

        assert_ne!(game.current_player, new_game.current_player);
    }

    #[test]
    fn has_dead_groups_black() {
        let mut game = GoBoard::empty();
        game.set_cell(
            BoardPosition::new(0, 0),
            BoardCell::Occupied(GoPlayer::Black),
        );
        game.set_cell(
            BoardPosition::new(0, 1),
            BoardCell::Occupied(GoPlayer::White),
        );
        game.set_cell(
            BoardPosition::new(1, 0),
            BoardCell::Occupied(GoPlayer::White),
        );

        assert!(game.has_dead_groups());
    }

    #[test]
    fn has_dead_groups_white() {
        let mut game = GoBoard::empty();
        game.set_cell(
            BoardPosition::new(0, 0),
            BoardCell::Occupied(GoPlayer::White),
        );
        game.set_cell(
            BoardPosition::new(0, 1),
            BoardCell::Occupied(GoPlayer::Black),
        );
        game.set_cell(
            BoardPosition::new(1, 0),
            BoardCell::Occupied(GoPlayer::Black),
        );

        assert!(game.has_dead_groups());
    }

    #[test]
    fn has_dead_groups_false() {
        let mut game = GoBoard::empty();
        game.set_cell(
            BoardPosition::new(0, 1),
            BoardCell::Occupied(GoPlayer::White),
        );
        game.set_cell(
            BoardPosition::new(1, 0),
            BoardCell::Occupied(GoPlayer::White),
        );

        assert!(!game.has_dead_groups());
    }

    #[test]
    fn hashing_is_stable() {
        assert_eq!(GoBoard::empty().stable_hash(), 13284472273662876477);
    }
}
