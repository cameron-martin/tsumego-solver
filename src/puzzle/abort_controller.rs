use std::time::{Duration, Instant};

pub trait AbortController {
    fn should_abort(&self) -> bool;
}

pub struct NoAbortController;

impl AbortController for NoAbortController {
    fn should_abort(&self) -> bool {
        false
    }
}

pub struct TimeoutAbortController {
    timeout_at: Instant,
}

impl AbortController for TimeoutAbortController {
    fn should_abort(&self) -> bool {
        Instant::now() >= self.timeout_at
    }
}

impl TimeoutAbortController {
    pub fn duration(duration: Duration) -> Self {
        TimeoutAbortController {
            timeout_at: Instant::now() + duration,
        }
    }
}
