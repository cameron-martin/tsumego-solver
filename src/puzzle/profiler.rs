use std::fmt;

pub trait Profiler {
    fn new() -> Self;
    fn move_up(&mut self);
    fn move_down(&mut self);
    fn visit_node(&mut self);
    fn order_success(&mut self);
    fn order_miss(&mut self);
}

pub struct NoProfile;

impl Profiler for NoProfile {
    fn new() -> NoProfile {
        NoProfile
    }

    fn move_up(&mut self) {}

    fn move_down(&mut self) {}

    fn visit_node(&mut self) {}

    fn order_success(&mut self) {}
    fn order_miss(&mut self) {}
}

pub struct Profile {
    current_depth: u8,
    pub max_depth: u8,
    pub visited_nodes: u32,
    successful_orderings: u32,
    missed_orderings: u32,
}

impl Profile {
    pub fn ordering_accuracy(&self) -> f32 {
        self.successful_orderings as f32
            / (self.successful_orderings + self.missed_orderings) as f32
    }
}

impl fmt::Display for Profile {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "Max Depth: {}\nVisited Nodes: {}\nOrdering Accuracy: {}\n",
            self.max_depth,
            self.visited_nodes,
            self.ordering_accuracy()
        )
    }
}

impl Profiler for Profile {
    fn new() -> Profile {
        Profile {
            current_depth: 1,
            max_depth: 1,
            visited_nodes: 0,
            successful_orderings: 0,
            missed_orderings: 0,
        }
    }

    fn move_up(&mut self) {
        self.current_depth -= 1;
    }

    fn move_down(&mut self) {
        self.current_depth += 1;
        if self.current_depth > self.max_depth {
            self.max_depth = self.current_depth;
        }
    }

    fn order_success(&mut self) {
        self.successful_orderings += 1;
    }

    fn order_miss(&mut self) {
        self.missed_orderings += 1;
    }

    fn visit_node(&mut self) {
        self.visited_nodes += 1;
    }
}
