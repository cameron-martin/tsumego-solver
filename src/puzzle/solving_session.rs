use super::{
    abort_controller::AbortController, example_collector::ExampleCollector,
    move_ranker::MoveRanker, solving_iteration::SolvingIteration, Profiler, Puzzle,
};
use crate::go::GoGame;
use std::{collections::HashSet, rc::Rc};

pub struct SolvingSession<'e, C: AbortController, P: Profiler, E: ExampleCollector, R: MoveRanker> {
    pub puzzle: Puzzle,
    pub move_ranker: Rc<R>,
    pub parents: HashSet<GoGame>,
    pub profiler: P,
    pub example_collector: &'e mut E,
    pub abort_controller: C,
}

impl<'e, C: AbortController, P: Profiler, E: ExampleCollector, R: MoveRanker>
    SolvingSession<'e, C, P, E, R>
{
    pub fn new(
        puzzle: Puzzle,
        abort_controller: C,
        example_collector: &'e mut E,
        move_ranker: Rc<R>,
    ) -> SolvingSession<C, P, E, R> {
        SolvingSession {
            parents: HashSet::new(),
            profiler: P::new(),
            example_collector,
            puzzle,
            abort_controller,
            move_ranker,
        }
    }

    pub fn create_iteration<'s>(
        &'s mut self,
        max_depth: u8,
    ) -> SolvingIteration<'s, 'e, C, P, E, R> {
        SolvingIteration::new(max_depth, self)
    }
}
