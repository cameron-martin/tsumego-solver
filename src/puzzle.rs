mod abort_controller;
mod profiler;
mod solution;
mod solving_iteration;
mod solving_session;
mod terminal_detection;

use crate::go::{GoGame, GoPlayer};
use abort_controller::{AbortController, NoAbortController, TimeoutAbortController};
pub use profiler::{NoProfile, Profile, Profiler};
pub use solution::Solution;
use solving_session::SolvingSession;
use std::time::Duration;

#[derive(Copy, Clone)]
pub struct Puzzle {
    pub game: GoGame,
    pub player: GoPlayer,
    pub attacker: GoPlayer,
}

impl Puzzle {
    pub fn new(game: GoGame) -> Puzzle {
        let attacker = if !(game.board.out_of_bounds().expand_one()
            & game.board.get_bitboard_for_player(GoPlayer::White))
        .is_empty()
        {
            GoPlayer::White
        } else {
            GoPlayer::Black
        };

        let player = game.current_player;

        Puzzle {
            game,
            player,
            attacker,
        }
    }

    pub fn from_sgf(sgf_string: &str) -> Puzzle {
        Self::new(GoGame::from_sgf(sgf_string))
    }

    pub fn solve<P: Profiler>(&self) -> Solution<P> {
        self.solve_with_controller::<_, P>(NoAbortController)
            .unwrap()
    }

    pub fn solve_with_timeout<P: Profiler>(&self, timeout: Duration) -> Option<Solution<P>> {
        self.solve_with_controller::<_, P>(TimeoutAbortController::duration(timeout))
    }

    fn solve_with_controller<C: AbortController, P: Profiler>(
        &self,
        abort_controller: C,
    ) -> Option<Solution<P>> {
        let mut max_depth: u8 = 1;

        let mut session = SolvingSession::new(*self, abort_controller);

        loop {
            let mut iteration = session.create_iteration(max_depth);
            let result = iteration.solve()?;

            if result != 0 {
                return Some(Solution {
                    won: result > 0,
                    principle_variation: iteration.principle_variation(),
                    profiler: session.profiler,
                });
            }

            max_depth += 1;
            session.profiler.move_down();
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::go::GoGame;
    use insta::{assert_display_snapshot, assert_snapshot};
    use profiler::Profile;
    use std::borrow::Borrow;

    fn show_principle_variation<P: Profiler>(puzzle: &Puzzle, solution: &Solution<P>) -> String {
        let games = solution
            .principle_variation
            .iter()
            .scan(puzzle.game, |state, &go_move| {
                *state = state.play_move(go_move).unwrap();

                Some((*state, go_move))
            });

        let mut output = String::new();
        output.push_str(&format!("{}\n\n", puzzle.game.board));
        for (game, go_move) in games {
            output.push_str(
                format!(
                    "{}: {}\n{}\n\n",
                    game.current_player.flip(),
                    go_move,
                    game.board
                )
                .borrow(),
            );
        }

        output
    }

    #[test]
    fn true_simple1() {
        let tsumego = GoGame::from_sgf(include_str!("test_sgfs/puzzles/true_simple1.sgf"));

        let puzzle = Puzzle::new(tsumego);

        let solution = puzzle.solve::<Profile>();

        assert!(solution.won);
        assert_display_snapshot!(solution.profiler.visited_nodes, @"644");
        assert_display_snapshot!(solution.profiler.max_depth, @"5");
        assert_snapshot!(show_principle_variation(&puzzle, &solution));
    }

    #[test]
    fn true_simple2() {
        let tsumego = GoGame::from_sgf(include_str!("test_sgfs/puzzles/true_simple2.sgf"));

        let puzzle = Puzzle::new(tsumego);

        let solution = puzzle.solve::<Profile>();

        assert!(solution.won);
        assert_display_snapshot!(solution.profiler.visited_nodes, @"1657");
        assert_display_snapshot!(solution.profiler.max_depth, @"7");
        assert_snapshot!(show_principle_variation(&puzzle, &solution));
    }

    #[test]
    fn true_simple3() {
        let tsumego = GoGame::from_sgf(include_str!("test_sgfs/puzzles/true_simple3.sgf"));

        let puzzle = Puzzle::new(tsumego);

        let solution = puzzle.solve::<Profile>();

        assert!(solution.won);
        assert_display_snapshot!(solution.profiler.visited_nodes, @"1490");
        assert_display_snapshot!(solution.profiler.max_depth, @"8");
        assert_snapshot!(show_principle_variation(&puzzle, &solution));
    }

    #[test]
    fn true_simple4() {
        let tsumego = GoGame::from_sgf(include_str!("test_sgfs/puzzles/true_simple4.sgf"));

        let puzzle = Puzzle::new(tsumego);

        let solution = puzzle.solve::<Profile>();

        assert!(solution.won);
        assert_display_snapshot!(solution.profiler.visited_nodes, @"31143");
        assert_display_snapshot!(solution.profiler.max_depth, @"7");
        assert_snapshot!(show_principle_variation(&puzzle, &solution));
    }

    // #[test]
    // fn true_medium1() {
    //     let tsumego = GoGame::from_sgf(include_str!("test_sgfs/puzzles/true_medium1.sgf"));

    //     let mut puzzle = Puzzle::<Profile>::new(tsumego);

    //     let solution = puzzle.solve();

    //     assert!(solution.won);
    //     // assert_eq!(puzzle.first_move(), Move::Place(BoardPosition::new(14, 2)));
    //     assert_display_snapshot!(puzzle.profiler.visited_nodes, @"345725");
    //     assert_display_snapshot!(puzzle.profiler.max_depth, @"29");
    // }

    #[test]
    fn true_ultrasimple1() {
        let tsumego = GoGame::from_sgf(include_str!("test_sgfs/puzzles/true_ultrasimple1.sgf"));

        let puzzle = Puzzle::new(tsumego);

        let solution = puzzle.solve::<Profile>();

        assert!(solution.won);
        assert_display_snapshot!(solution.profiler.visited_nodes, @"5");
        assert_display_snapshot!(solution.profiler.max_depth, @"1");
        assert_snapshot!(show_principle_variation(&puzzle, &solution));
    }

    #[test]
    fn true_ultrasimple2() {
        let tsumego = GoGame::from_sgf(include_str!("test_sgfs/puzzles/true_ultrasimple2.sgf"));

        let puzzle = Puzzle::new(tsumego);

        let solution = puzzle.solve::<Profile>();

        assert!(solution.won);
        assert_display_snapshot!(solution.profiler.visited_nodes, @"1939");
        assert_display_snapshot!(solution.profiler.max_depth, @"8");
        assert_snapshot!(show_principle_variation(&puzzle, &solution));
    }
}
