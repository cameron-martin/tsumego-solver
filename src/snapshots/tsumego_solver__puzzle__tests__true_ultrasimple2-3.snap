---
source: src/puzzle.rs
expression: "show_principle_variation(&puzzle, &solution)"
---
. . . w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _


black: (1, 0)
. b . w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _


white: Pass
. b . w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _


black: (0, 0)
b b . w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _


white: (2, 0)
. . w w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _


black: Pass
. . w w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _


white: (0, 0)
w . w w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _


black: Pass
w . w w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _


white: Pass
w . w w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _



